### PR TITLE
Add toil-lib dependency (resolves BD2KGenomics/toil#922)

### PR DESCRIPTION
**Pull request by jvivian**

### CODE REVIEW

**Comment by fnothaft** [comment]
I like the +/- spread on this! Reviewing now.

File: setup.py [hunk 1]
```diff
@@ -73,6 +73,7 @@ def check_provided(distribution, min_version, max_version=None, optional=False):
     author_email='cgl-toil@googlegroups.com',
     url="https://github.com/BD2KGenomics/toil-scripts",
     install_requires=[
+        'toil-lib==1.0.2',
         'tqdm==3.8.0',  # FIXME: Remove once ADAM stops using it (superfluous import)
```
Comment (fnothaft) [inline setup.py]: Small aside, but I think we can get rid of the `tqdm` requirement. I don't see it used anywhere in the codebase.

File: src/toil_scripts/adam_gatk_pipeline/align_and_call.py [hunk 2]
```diff
@@ -113,11 +113,10 @@
 
 # import from python system libraries
 import argparse
-import copy
 import textwrap
+import copy
```
Comment (fnothaft) [inline src/toil_scripts/adam_gatk_pipeline/align_and_call.py]: Lex sort imports; this one shouldn't have moved.

File: setup.py (hunk 1 above)
Comment (jvivian) [inline setup.py]: That would be great

File: src/toil_scripts/adam_gatk_pipeline/align_and_call.py [hunk 3]
```diff
@@ -130,7 +129,7 @@
 from toil_scripts.gatk_processing.gatk_preprocessing import * #download_gatk_files
 from toil_scripts.rnaseq_cgl.rnaseq_cgl_pipeline import generate_file
 
-from toil_scripts.lib.programs import mock_mode
+from toil_lib.programs import mock_mode
```
Comment (fnothaft) [inline src/toil_scripts/adam_gatk_pipeline/align_and_call.py]: `toil_lib` imports should be ahead of `toil_scripts` imports.

File: src/toil_scripts/adam_pipeline/adam_preprocessing.py [hunk 4]
```diff
@@ -39,11 +39,11 @@
...
 
-from toil_scripts.lib import require
-from toil_scripts.lib.files import copy_files, move_files
-from toil_scripts.lib.programs import docker_call, mock_mode
+from toil_lib import require
+from toil_lib.files import copy_files, move_files
+from toil_lib.programs import docker_call, mock_mode
 from toil_scripts.rnaseq_cgl.rnaseq_cgl_pipeline import generate_file
-from toil_scripts.tools.spark_tools import call_adam, call_conductor, MasterAddress, HDFS_MASTER_PORT, SPARK_MASTER_PORT
+from toil_lib.tools.spark_tools import call_adam, call_conductor, MasterAddress, HDFS_MASTER_PORT, SPARK_MASTER_PORT
```
Comment (fnothaft) [inline src/toil_scripts/adam_pipeline/adam_preprocessing.py]: `toil_lib` imports should be ahead of `toil_scripts` imports.

File: src/toil_scripts/gatk_germline/germline.py [hunk 5]
```diff
@@ -42,10 +42,11 @@
 
 import yaml
 from toil.job import Job
+from toil_lib import require
```
Comment (fnothaft) [inline src/toil_scripts/gatk_germline/germline.py]: 1 line of whitespace between `toil` and `toil_lib` imports? (I thought that was the policy, but can't remember)

File: src/toil_scripts/gatk_processing/gatk_preprocessing.py [hunk 6]
```diff
@@ -34,9 +34,9 @@
 import yaml
 from toil.job import Job
 from toil_scripts import download_from_s3_url
-from toil_scripts.lib import require
-from toil_scripts.lib.programs import docker_call
-from toil_scripts.lib.urls import s3am_upload
+from toil_lib import require
```
Comment (fnothaft) [inline src/toil_scripts/gatk_processing/gatk_preprocessing.py]: `toil_lib` imports should be ahead of `toil_scripts` imports.

File: src/toil_scripts/rnaseq_unc/rnaseq_unc_pipeline.py [hunk 7]
```diff
@@ -55,17 +55,18 @@
...
 import hashlib
-import os
-import subprocess
-import errno
 import multiprocessing
+import os
 import shutil
+import subprocess
 import tarfile
+from collections import OrderedDict
```
Comment (fnothaft) [inline src/toil_scripts/rnaseq_unc/rnaseq_unc_pipeline.py]: I don't think we were moving the `from` imports to the end.

File: src/toil_scripts/spladder_pipeline/spladder_pipeline.py [hunk 8]
```diff
@@ -29,16 +29,15 @@
...
-
-from toil_scripts.lib.files import tarball_files
-from toil_scripts.lib.jobs import map_job
-from toil_scripts.lib.programs import docker_call
-from toil_scripts.lib.urls import s3am_upload_job
-from toil_scripts.lib.urls import download_url_job
-from toil_scripts.lib.urls import download_url
-from bd2k.util.processes import which
 from bd2k.util.files import mkdir_p
+from bd2k.util.processes import which
```
Comment (fnothaft) [inline src/toil_scripts/spladder_pipeline/spladder_pipeline.py]: Space between `bd2k` and `toil` and `toil_lib` imports?

File: setup.py (hunk 1 above)
Comment (fnothaft) [inline setup.py]: Let's just remove it in this PR?

**Comment by fnothaft** [comment]
Just dropped a few nits. Other than nits, LGTM!

File: src/toil_scripts/gatk_germline/germline.py (hunk 5 above)
Comment (jvivian) [inline src/toil_scripts/gatk_germline/germline.py]: I ran optimize imports before realizing it messed up the `align_and-call` pipeline's imports, so I just did a quick find-and-replace. I'll go back and do that.

**Comment by jvivian** [comment]
Thanks @fnothaft! I'll address these now.

**Comment by jvivian** [comment]
@fnothaft — found a blunder in toil-lib. I'll need to push 1.0.3 out before I can merge this, doing that now.

**Comment by fnothaft** [comment]
Blimey mate! A blunder!

File: src/toil_scripts/gatk_germline/germline.py (hunk 5 above)
Comment (jvivian) [inline src/toil_scripts/gatk_germline/germline.py]: PyCharm's `optimize imports` feature doesn't put a space there, but puts 1 line of whitespace between that block and toil_scripts.  Do you have a preference?

**Comment by fnothaft** [comment]
@jvivian so now that this is merged, PRs that added to toil_scripts.tools or toil_scripts.lib should be split, right? E.g., https://github.com/BD2KGenomics/toil-scripts/pull/381 added to `toil_scripts.tools.qc`. Is the new workflow that we open a PR against `toil-lib`, get that merged, bump the `toil-lib` version, and then PR the rest here? Or, actually, should new functionality go into it's own repository? E.g., should #381 be split into a PR against `toil-lib` and a new repository?

**Comment by jvivian** [comment]
That's a good question.  I think the former option is what I'll do in the short term until a pipeline gets broken out into its own repo.  Once a pipeline has broken up with its estranged lover toil-scripts, PRs can target the new repo with changes that complement the additions to toil-lib. 

PR against toil-lib > merge > bump toil-lib to new release > build release branch on jenkins to push to pypi > bump toil-lib to dev version > PR against toil-scripts | PR against pipeline-repo.

**Comment by fnothaft** [comment]
SGTM. I'm not planning to touch #381 for a while, so I'll hold off there.

**Pull request merged**